### PR TITLE
Add `no-locs` compiler flag for usage in tests

### PR DESCRIPTION
**Pull request by jbachurski**
## Background

ocamltest's expect tests in the `testsuite/` often feature very long chains of unrelated tests, some of which raise (compiler) errors. When such errors are raised, they feature line numbers, as in:

```ocaml
<[ let o = object method f = 1 end in o#f ]>;;
[%%expect {|
Line 690, characters 11-34:
690 | <[ let o = object method f = 1 end in o#f ]>;;
                 ^^^^^^^^^^^^^^^^^^^^^^^
Error: Object definition using "object..end"
       is not supported inside quoted expressions,
       as seen at Line 690, characters 11-34.
|}];;
```

These line numbers are (usually) unhelpful and cause flakiness: e.g. promoting outputs after changing line numbers causes unhelpful diffs and instability under `-principal` runs.

_On that note, I find it extremely amusing that the CI for this PR initially failed because of line numbers changing in test outputs._

## Feature

This PR adds the flag `-no-locs`, which aims to disable printing of line numbers and column numbers within `print_loc` and `highlight_quote` in `location.ml`. These should be the main reasonable sites where line/column numbers are revealed.

The same test file as in the example above with `-no-locs` passed will now output:

```
<[ let o = object method f = 1 end in o#f ]>;;
[%%expect {|
Line _, characters _-_:
 | <[ let o = object method f = 1 end in o#f ]>;;
              ^^^^^^^^^^^^^^^^^^^^^^^
Error: Object definition using "object..end"
       is not supported inside quoted expressions,
       as seen at Line _, characters _-_.
|}];;
```

Note that locations are still (mostly) tested with the `^^^` markers - these markets are internally referred to as 'quote highlights'. Also note these are not present a location spans multiple lines.

As a preview, I enabled the new flags in `build_quotation.ml`.

## Use cases

Besides tests in the compiler, some tools will either explicitly sanitise test outputs for line/column numbers with regexes, while tools like [`toplevel_expect_tests`](https://github.com/janestreet/toplevel_expect_test/blob/afa7267f7cf5c0d82b68a92353428821b9897700/src/main.ml#L65) override the location printer using internal magic (though, thanks to this, locations can be enabled/disabled dynamically).  These use cases are subsumed by this pair of flags.

## Alternatives

This could be done as a regex-replace pass on `ocamltest`, but there's a bunch of possible formats in which line/column numbers occur (mixes of capitalisation and whether it is a single point or a range), so it feels better to fix this at its source with the possibility to use these flags in other places.

### CODE REVIEW

File: driver/main_args.ml
Comment (dkalinichenko-js) [inline driver/main_args.ml:1]: I think it makes more sense to only have a single flag that disables both line and column numbers. I can't imagine many situations where you'd want to disable one but not the other, and disabling only one of them still leaves the test fragile.

File: parsing/location.ml [hunk 1]
```diff
@@ -565,9 +579,12 @@ let highlight_quote ppf
         |> List.map (fun ({ text; start_pos } as line) ->
           let end_pos = start_pos + String.length text - 1 in
           let line_nb =
-            match ISet.find_bound_in iset ~range:(start_pos, end_pos) with
-            | None -> None
-            | Some (p, _) -> Some p.pos_lnum
+            match
+              ISet.find_bound_in iset ~range:(start_pos, end_pos),
```
Comment (dkalinichenko-js) [inline parsing/location.ml:583]: Why the change here? What difference in behavior does it lead to?

File: testsuite/tests/quotation/build_quotation.ml
Comment (dkalinichenko-js) [inline testsuite/tests/quotation/build_quotation.ml:1]: I think it would be better to test this in a separate new file and try to test some edge cases (like a nonexistent location). Probably worth having both an expect test and a normal compiler output test.

File: driver/main_args.ml
Comment (jbachurski) [reply driver/main_args.ml:1]: Hmm, my motivation was that `toplevel_expect_tests` allowed changing either, and because column numbers are less fragile than line numbers. 
I'm happy with a `-no-locs` / `-no-locations` / `-suppress-locations` flag if you think that makes more sense - should I get rid of the fine-graining in the internal logic too?

File: parsing/location.ml (hunk 1 above)
Comment (jbachurski) [reply parsing/location.ml:583]: This corresponds to the line preview part -- the `690 | ...` in the example in the description. I believe this bit is referred to in `Location` as a 'quote'.
In particular it's a red flag that the line number is printed, so we want to hide it (None just turns it into a blank space, consistent with an "unknown" location).

File: testsuite/tests/quotation/build_quotation.ml
Comment (jbachurski) [reply testsuite/tests/quotation/build_quotation.ml:1]: I was going to but couldn't find a test directory that tests compiler flags like this -- any suggestions?
(this is just bikeshedding over a name, but I'd rather not make the existing naming scheme even more inconsistent)

By the way, how do you trigger a non-existent location?

File: driver/main_args.ml
Comment (dkalinichenko-js) [reply driver/main_args.ml:1]: Yep, I don't think it's worth having.

File: parsing/location.ml (hunk 1 above)
Comment (dkalinichenko-js) [reply parsing/location.ml:583]: Ah, I see.

File: testsuite/tests/quotation/build_quotation.ml
Comment (dkalinichenko-js) [reply testsuite/tests/quotation/build_quotation.ml:1]: `testsuite/tests/messages` looks good enough.

Search for tests printing `File "_none_"` to see what you can use as an example.

File: parsing/location.ml (hunk 1 above)
Comment (dkalinichenko-js) [reply parsing/location.ml:583]: In this case, best to demonstrate the change in a test.

File: driver/main_args.ml
Comment (dkalinichenko-js) [reply driver/main_args.ml:1]: `-no-locs` is the best option here, imo, no reason to add extra length.

I think it's also worth having a `-locs` flag to override that, so `ocamltest` can set `-no-locs` by default while we re-enable locations where we need them.
